packages/llamaindex/src/tools/AzureDynamicSessionTool.ts

```diff
@@ -0,0 +1,355 @@
+
+  /**
+   * The path to save the downloaded file to.
+   * If not provided, the file is returned as a ReadableStream.
+   * @example "/path/to/file.txt"
+   */
+  localFilename?: string;
+};
+
+type UploadFileMetadata = {
+  /**
+   * The data to upload
+   */
+  data: Buffer;
+
+  /**
+   * The path to the local file to upload
+   * @example "file.txt"
+   * @example "folder/file.txt"
+   */
+  remoteFilename: string;
+};
+
+let _userAgent = "";
+
+/**
+ * A utility function to generate the user agent in the format:
+ *
+ * `llamaIndex-azure-dynamic-sessions (Language=TypeScript; node.js/v14.17.0; darwin/x64)`
+ * @returns The user agent string.
+ */
+async function getuserAgentSuffix(): Promise<string> {
+  try {
+    //@ts-ignore
+    const __filename = fileURLToPath(import.meta.url);
+    const __dirname = dirname(__filename);
+
+    if (!_userAgent) {
+      const data = await readFile(
+        join(__dirname, "..", "package.json"),
+        "utf8",
+      );
+      const json = await JSON.parse(data.toString());
+      _userAgent = `${json.name}/${json.version}`;
+    }
+  } catch (e) {
+    _userAgent = `llamaIndex-azure-dynamic-sessions`;
+  }
+  return `${_userAgent} (Language=TypeScript; node.js/${process.version}; ${process.platform}; ${process.arch})`;
+}
+
+function getAzureADTokenProvider() {
+  return getBearerTokenProvider(
+    new DefaultAzureCredential(),
+    "https://dynamicsessions.io/.default",
+  );
+}
+
+const DEFAULT_META_DATA: ToolMetadata = {
+  name: "azure_dynamic_sessions_python_interpreter",
+  description:
+    "A Python shell. Use this to execute python commands " +
+    "when you need to perform calculations or computations. " +
+    "Input should be a valid python command. " +
+    "Returns the result, stdout, and stderr. ",
+  parameters: {
+    type: "object",
+    properties: {
+      code: {
+        type: "string",
+        description: "The Python code to execute",
+      },
+    },
+    required: ["code"],
+  },
+};
+
+/**
+ * Azure Code Interpreter tool: A tool that allows you to interact with a dynamic session on Azure.
+ */
+export class AzureDynamicSessionTool
+  implements BaseTool<AzureDynamicSessionToolParams>
+{
+  /**
+   * The metadata for the tool.
+   */
+  metadata: ToolMetadata;
+
+  /**
+   * The session ID to use for the session pool. Defaults to a random UUID.
+   */
+  sessionId: string;
+
+  /**
+   * The endpoint of the Azure pool management service.
+   * This is where the tool will send requests to interact with the session pool.
+   * If not provided, the tool will use the value of the `AZURE_CONTAINER_APP_SESSION_POOL_MANAGEMENT_ENDPOINT` environment variable.
+   */
+  poolManagementEndpoint: string;
+
+  /**
+   * A function that returns the access token to use for the session pool.
+   */
+  private azureADTokenProvider: () => Promise<string>;
+
+  constructor(params?: AzureDynamicSessionToolParams) {
+    this.metadata = params?.metadata || DEFAULT_META_DATA;
```

Comment:
can we define a output directory where the tool stores generate files, e.g. images? similar to https://github.com/run-llama/create-llama/blob/293557cbb4b4717ee18e3ed27f92ebdc53c0bf49/templates/components/engines/typescript/agent/tools/interpreter.ts#L59

Reply:
Defintely. Will add that.

Reply:
done.